### PR TITLE
Undertale: Fixes a major logic bug, and updates Undertale to use the new Options API

### PR DESCRIPTION
**Pull request by jonloveslegos**
## What is this fixing or adding?

Fixes a logic bug with Undertale that results in the potential of a near post-game location potentially being required prior to being able to reach it.

Updates the Undertale options.py to use the new Options API.

## How was this tested?

This was tested by successfully generating a seed with Undertale, and through successfully starting and connecting the client.

### CODE REVIEW

File: worlds/undertale/Rules.py [hunk 1]
```diff
@@ -101,7 +101,10 @@ def set_rules(multiworld: MultiWorld, player: int):
         set_rule(multiworld.get_location("True Lab Plot", player),
                  lambda state: state.can_reach("New Home", "Region", player)
                                and state.can_reach("Letter Quest", "Location", player)
-                               and state.can_reach("Alphys Date", "Location", player))
+                               and state.can_reach("Alphys Date", "Location", player)
+                               and (state.has("Left Home Key", player) and
+                                    state.has("Right Home Key", player)) or
```
Comment (Exempt-Medic) [inline worlds/undertale/Rules.py:106]: Fairly certain this paren is supposed to go after the next line
```suggestion
                                    state.has("Right Home Key", player) or
```

File: worlds/undertale/Rules.py [hunk 2]
```diff
@@ -101,7 +101,10 @@ def set_rules(multiworld: MultiWorld, player: int):
         set_rule(multiworld.get_location("True Lab Plot", player),
                  lambda state: state.can_reach("New Home", "Region", player)
                                and state.can_reach("Letter Quest", "Location", player)
-                               and state.can_reach("Alphys Date", "Location", player))
+                               and state.can_reach("Alphys Date", "Location", player)
+                               and (state.has("Left Home Key", player) and
```
Comment (Exempt-Medic) [inline worlds/undertale/Rules.py:105]: Fairly certain this is supposed to be like this:
```py
                               and (state.has("Left Home Key", player) and
                                    state.has("Right Home Key", player) or
                                    state.has("Key Piece", player, state.multiworld.key_pieces[player].value)))
```

**Comment by Exempt-Medic** [comment]
While you're doing this, might as well change your use of `per_slot_randoms` in init as well

**Comment by Exempt-Medic** [comment]
You should also update `Rules.py`

**Comment by jonloveslegos** [comment]
> While you're doing this, might as well change your use of `per_slot_randoms` in init as well

Hopefully I did that right? I don't remember what exactly I was suppose to change it to, so I looked at dlcquest

**Review by Exempt-Medic — APPROVED** [review]
Changes LGTM. Couldn't hit or find any deprecated option getters, made a plando seed that hit the fixed logic bug on the new code and worked anyway on the old code, the Shop 4 item was no longer listed as inaccessible with Temmie Flake replacing every item. Generations before and after that didn't hit these changes were the same. There are some lines that could be changed, like `bool(world.options.rando_love.value)` but I'm just gonna let core decide whether they care about that

File: worlds/undertale/__init__.py [hunk 3]
```diff
@@ -46,47 +46,64 @@ class UndertaleWorld(World):
...
+        elif self.options.route_required == "genocide":
+            junk_pool = junk_weights_genocide.copy()
+        elif self.options.route_required == "neutral":
+            junk_pool = junk_weights_neutral.copy()
+        elif self.options.route_required == "pacifist":
+            junk_pool = junk_weights_pacifist.copy()
+        else:
+            junk_pool = junk_weights_all.copy()
+        if not self.options.only_flakes:
+            return self.random.choices(list(junk_pool.keys()), weights=list(junk_pool.values()))[0]
```
Comment (NewSoupVi) [inline worlds/undertale/__init__.py:90]: Hmmm I'm pretty sure that at no point during code does `junk_pool` get modified, so I don't understand why you are making a `.copy()` of the respective junk weights

File: worlds/undertale/__init__.py [hunk 4]
```diff
@@ -99,110 +116,107 @@ def create_items(self):
...
             itempool = [item for item in itempool if item != "Popato Chisps" and item != "Stained Apron" and
                         item != "Nice Cream" and item != "Hot Cat" and item != "Hot Dog...?" and item != "Punch Card"]
-        elif self.multiworld.route_required[self.player] == "neutral":
+        elif self.options.route_required == "neutral":
             itempool = [item for item in itempool if item != "Popato Chisps" and item != "Hot Cat" and
                         item != "Hot Dog...?"]
-        if self.multiworld.route_required[self.player] == "pacifist" or \
-                self.multiworld.route_required[self.player] == "all_routes":
+        if self.options.route_required == "pacifist" or \
+                self.options.route_required == "all_routes":
```
Comment (NewSoupVi) [inline worlds/undertale/__init__.py:133]: Does this line still need the `\` now?

File: worlds/undertale/__init__.py [hunk 5]
```diff
@@ -99,110 +116,107 @@ def create_items(self):
...
+        if not self.options.rando_love or \
+                (self.options.route_required != "genocide" and
+                 self.options.route_required != "all_routes"):
             itempool = [item for item in itempool if not item == "LOVE"]
-        if not self.multiworld.rando_stats[self.player] or \
-                (self.multiworld.route_required[self.player] != "genocide" and
-                 self.multiworld.route_required[self.player] != "all_routes"):
+        if not self.options.rando_stats or \
+                (self.options.route_required != "genocide" and
+                 self.options.route_required != "all_routes"):
```
Comment (NewSoupVi) [inline worlds/undertale/__init__.py:151]: Same thing here, wonder if these could be reformatted to look nicer now

File: worlds/undertale/__init__.py [hunk 6]
```diff
@@ -99,110 +116,107 @@ def create_items(self):
...
+        if not self.options.rando_love or \
+                (self.options.route_required != "genocide" and
+                 self.options.route_required != "all_routes"):
             exclusion_pool.update(exclusion_table["NoLove"])
-        if not self.multiworld.rando_stats[self.player] or \
-                (self.multiworld.route_required[self.player] != "genocide" and
-                 self.multiworld.route_required[self.player] != "all_routes"):
+        if not self.options.rando_stats or \
+                (self.options.route_required != "genocide" and
+                 self.options.route_required != "all_routes"):
```
Comment (NewSoupVi) [inline worlds/undertale/__init__.py:185]: see above

File: worlds/undertale/__init__.py [hunk 7]
```diff
@@ -99,110 +116,107 @@ def create_items(self):
...
-        if not self.multiworld.only_flakes[self.player]:
-            itempool += self.multiworld.random.choices(list(junk_pool.keys()), weights=list(junk_pool.values()),
-                                                       k=len(self.location_names)-len(itempool)-len(exclusion_pool))
-        else:
-            itempool += ["Temmie Flakes"] * (len(self.location_names) - len(itempool) - len(exclusion_pool))
         # Convert itempool into real items
         itempool = [item for item in map(lambda name: self.create_item(name), itempool)]
+        # Fill remaining items with randomly generated junk or Temmie Flakes
+        while len(itempool) < len(self.multiworld.get_unfilled_locations(self.player)):
+            itempool += [self.create_filler()]
```
Comment (NewSoupVi) [inline worlds/undertale/__init__.py:197]: **Extreme nitpick alert**

`itempool.append(self.create_filler())` is faster

*Low Priority*

**Review by NewSoupVi — CHANGES_REQUESTED** [review]
I have some nitpicks but other than that this is good to go

**Review by agilbert1412 — APPROVED** [review]
I agree with Violet's suggestions, and have no more to offer. Approved, assuming these get addressed

File: worlds/undertale/__init__.py (hunk 5 above)
Comment (jonloveslegos) [reply worlds/undertale/__init__.py]: These lines could only have the line separation removed for the route_required parentheses section

**Review by NewSoupVi — APPROVED** [review]

**Pull request merged**